### PR TITLE
Refactor Action implementation

### PR DESCRIPTION
**Pull request by lundibundi**
Add collection domain

* make Action an object with properties
* add Execute decorator to denote Action reference
* add type, transactionMode, processMode, refreshMode props
* allow to specify custom Form name via Form field

### CODE REVIEW

**Comment by tshemsedinov** [comment]
FYI: @aqrln we need to know context specification (schema) to understand how to implement this PR better. Example (from your paper):
```js
context.schema: {
  Id: { domain: 'Id', required: false, comment: 'current record id' },
  SelectedIds: { domain: 'List', required: false, comment: 'records selected in ui' },
  Parent: { domain: 'Id', required: false, comment: 'direct parent id' },
  Master: { domain: 'Id', required: false, comment: 'high-level id' },
  Category: { domain: 'Nomen', required: true, comment: 'category name' },
  ParentCategory: { domain: 'Nomen', required: false },
  MasterCategory: { domain: 'Nomen', required: false },
  ProcessId: { required: false, comment: 'business-process id' },
  DisplayMode: { domain: 'Nomen', required: false, comment: 'old name show method' },
  DisplayModeParameters: { domain: 'Collection', required: false },
}
```

File: lib/decorators.js [hunk 1]
```diff
@@ -78,7 +78,30 @@ class Unique extends Index {}
...
+}
+
+class Action {
+  constructor(def) {
+    if (typeof def === 'function') {
+      this.Execute = def;
+    } else {
+      Object.assign(this, def);
+    }
+    if (!this.type) this.type = 'Custom';
```
Comment (tshemsedinov) [inline lib/decorators.js]:
```suggestion
    if (!this.Type) this.Type = 'Custom';
```

File: lib/decorators.js [hunk 2]
```diff
@@ -78,7 +78,30 @@ class Unique extends Index {}
...
+    if (!this.transactionMode) this.transactionMode = 'All';
```
Comment (tshemsedinov) [inline lib/decorators.js]:
```suggestion
    if (!this.TransactionMode) this.TransactionMode = 'All';
```

File: lib/decorators.js [hunk 3]
```diff
@@ -78,7 +78,30 @@ class Unique extends Index {}
...
+    if (!this.processMode) this.processMode = 'Any';
```
Comment (tshemsedinov) [inline lib/decorators.js]:
```suggestion
    if (!this.ProcessMode) this.ProcessMode = 'Any';
```

File: lib/decorators.js [hunk 4]
```diff
@@ -78,7 +78,30 @@ class Unique extends Index {}
...
+    if (!this.refreshMode) this.refreshMode = 'None';
```
Comment (tshemsedinov) [inline lib/decorators.js]:
```suggestion
    if (!this.RefreshMode) this.RefreshMode = 'None';
```

File: schemas/metaschema/domains.schema [hunk 5]
```diff
@@ -66,4 +67,19 @@
   Volume: { type: 'number', unit: 'Litre', symbol: 'l' },
   Speed: { type: 'number', unit: 'Metres per second', symbol: 'm/s' },
   Ratio: { type: 'number' },
+  ActionType: Enum(
+    'Create', 'Update', 'Delete', 'ShowDetail',
+    'Digicert', 'Steps', 'Custom'
```
Comment (belochub) [inline schemas/metaschema/domains.schema]: What's the `Steps` action type, how will it behave?

Comment (lundibundi) [reply schemas/metaschema/domains.schema]: As of now it's not really specified how will `steps` work.
It describes an Action that may happen via multiple interactions with the user (and corresponding executions of different Actions) while keeping some internal state.

There is no use for it now so I can remove it, though we will probably need to add it later anyway.

Comment (belochub) [reply schemas/metaschema/domains.schema]: Doesn't this PR add an `Execute` decorator, that is supposed to determine the next step in the interaction with the user?
Or did I misunderstand the purpose of it and there is a significant difference?

Comment (lundibundi) [reply schemas/metaschema/domains.schema]: Yes, this is a basic idea - to return the next step via Execute, but that may not be enough when we actually get to implementing such an interaction.

Comment (tshemsedinov) [reply schemas/metaschema/domains.schema]: `'Create', 'Update', 'Delete'` is ok, but we don't need `'ShowDetail', 'Digicert', 'Steps', 'Custom'` @lundibundi

File: schemas/metaschema/domains.schema [hunk 6]
```diff
@@ -66,4 +67,19 @@
...
+  ),
+  // determines how Action with multiple items will be contained
+  // Each - commit after each item, All - commit after all
+  TransactionMode: Enum('Each', 'All'),
```
Comment (tshemsedinov) [inline schemas/metaschema/domains.schema]:
```suggestion
  TransactionMode: Enum('Single', 'All'),
```

Comment (tshemsedinov) [reply schemas/metaschema/domains.schema]: `All` by default. By the way if `Action` is `Sequential` or `Parallel` we will disallow to return `Execute` for multiple entities.

File: schemas/metaschema/domains.schema [hunk 7]
```diff
@@ -66,4 +67,19 @@
...
+  // determines how Action uses selected items
+  // Any - may use any elements, Current - uses only current element
+  // Selected - uses all user selected items
+  ProcessMode: Enum('Any', 'Current', 'Selected'),
+  // determines what rows to update after the Action
+  // None - don't update, Current - update only affected items
+  // All - update whole dataset
+  RefreshMode: Enum('None', 'Current', 'All'),
```
Comment (tshemsedinov) [inline schemas/metaschema/domains.schema]:
```suggestion
  RefreshMode: Enum('Current', 'All', 'None'), // Current by defeult
```

**Comment by lundibundi** [comment]
Landed in https://github.com/metarhia/metaschema/commit/b446ecc939523cf86c6fd0ab7ad9969edacb563f and https://github.com/metarhia/metaschema/commit/ed310873b3c78464405634720c52280ee8d288e3.

**Pull request closed without merging**